### ISSUE TITLE
please support XMP tag "subject"

### ISSUE THREAD

**Issue by kagahd** [bpatrik/pigallery2#267]
I discovered PiGallery2 only yesterday and I love it!
However, many of my photos store individual labels in the XMP section "subject". Unfortunately pigallery2 is unable to read (and search) such data.

I adapted PiGallery2 to fit my needs as follows:
At line 204 of file ./src/backend/model/threading/MetadataLoader.js I added the following code:

```
if(exif.subject && exif.subject.value){
   metadata.keywords = [];
   for(let i=0; i < exif.subject.value.length; i++){
      metadata.keywords.push(exif.subject.value[i].value);
   }
}
```
This works well for me but as you can see, my code **replaces** the "keywords" by my "subject" values. This is of course not a good solution for somebody who uses "keywords" (I don't use keywords). Hence my feature request to implement "subject" XMP metadata properly as a new property. 

Please excuse my ignorance to do it by myself. I think to implement this properly, at least the data model, DTO's and tests need to be adapted which I feel I'm not capable of. Thanks for your understanding and your effort!

**Comment by bpatrik** [comment]
Hi
Can you elaborate more about this subject XMP section? Why do you use it instead of the iptc keywords? Do they have different purpose? Are you using a tool that puts these tags to subject instead of the normal keywords section?

If it is a valid use, I do not mind adding it to the lib. We can merge the current keywords list with this list of subjects. 
If it is something that is not standard, I would prefer not adding to the code.

**Comment by kagahd** [comment]
Hi @bpatrik 
I'm using the Android app "F-Stop Gallery Pro" to tag my photos.

After a bit googling, what the "subject" field is for, I found out that _**"Subject" is the XMP place for keywords**._ ([source](https://www.carlseibert.com/xmp-iptciim-or-exif-which-is-preferred/)) 

The [specification](https://www.iptc.org/std/photometadata/specification/IPTC-PhotoMetadata#keywords) of the XMP tag  `dc:subject` confirmt this statement:

Row header | Specification
-- | --
Name  | Keywords
Definition | Keywords to express the subject of the image. Keywords may be free text and don’t have to be taken from a controlled vocabulary. Codes from the controlled vocabulary IPTC Subject NewsCodes must go to the "Subject Code" field.
Help Text | Enter any number of keywords, terms or phrases used to express the subject matter in the image.
Implementation Note(s) | Single values of this field should not be restricted to single words but must allow for phrases as well.
Label | Keywords
Basic Specs | Data type: Text / Cardinality: 0..unbounded
XMP Specs | dc:subject [Bag Text <External>]
IIM Specs | 2:25 Keywords
JSON Specs | keywords [string//array]

Obviously, the XMP `subject` field corresponds to the `keyword` field according to the IIM and JSON specification.

See also:
[8.2. Synchronising IIM elements with existing XMP properties](https://www.iptc.org/std/photometadata/specification/IPTC-PhotoMetadata#synchronising-iim-elements-with-existing-xmp-properties)

Photoshop Name | IIM Spec | XMP Spec
-- | -- | --
Keywords | 2:25 Keywords | dc:subject

So your suggestion would be great to merge the `subject` field to the keywords.
Thank you!

**Comment by bpatrik** [comment]
Thanks for the detailed explanation. 

I Think we can add xmp subject support (as normal keywords)

Would you have time to create the PR for that?
Can you please also add a test photo (pls keep photo size < 100kb) ? Something like this. https://github.com/bpatrik/pigallery2/blob/90a6c4b6b1d2a2e22411288ad5e2613fc62de38a/test/backend/unit/model/threading/MetaDataLoader.spec.ts#L30-L34

You can put it under assets/xmp (by also creating the xmp folder).

**Comment by kagahd** [comment]
Hi @bpatrik,

I just created the PR, including the photo and expected json.

**Comment by bpatrik** [comment]
Thanks, can you maybe also include the actual implementation in the PR too? :)
 So far it only makes the tests fail:
![kép](https://user-images.githubusercontent.com/7512182/115919586-53de2e80-a479-11eb-8341-f57e3f3d3cce.png)
link: https://github.com/bpatrik/pigallery2/pull/270/checks?check_run_id=2422510897
Also pls add all info that the metadata loader can extract from the photo to the json file

**Comment by kagahd** [comment]
I included the implementation in order to merge the XMP 'subject' section to 'keywords'. I also added the meta info that the test expected. However, the test now fails by saying that it expects other fields of other photos that I did not touch. I don't know what to do in this case.

**Comment by kagahd** [comment]
I found out that some photos have both IPTC `keywords` and XMP `subject` tags. However, my code **added** the XMP `subject` tags to the `keywords` which was not expected by the tests. 
To solve this, my code now adds the XMP `subject` tags only if its value is not yet in the `keywords`. However, the tests still fail for the photo having special characters, the photo with faces and my photo. I don't understand why the test expects [theses](https://travis-ci.org/github/bpatrik/pigallery2/jobs/768217454#L1023) entries because I don't see them in XnView either.
I further don't understand why the test with my photo could neither read its [rating](https://travis-ci.org/github/bpatrik/pigallery2/jobs/768217454#L1098) nor its [keywords](https://travis-ci.org/github/bpatrik/pigallery2/jobs/768217454#L1108). Or am I misinterpreting the output of the test?
Please advise. Any help is greatly appreciated.

**Comment by kagahd** [comment]
The tests of the last [commit](https://github.com/kagahd/pigallery2/commit/f8b68e5c2c0cbdb13dbbd1ce2acfc55a3e670c79) are all green. 

For completeness, I describe the reasons why the tests of my former commits failed: 

My modification reads more metadata from the files while the tests did not reflected them. That's the case for the `test_png.png` file which has 2 keywords in the XMP `subject` section. However, I had to add `"keywords":[ ]` to the `test_png.json` because the keywords were **not** retrieved by the `ExifReader`. This might be worth a new issue. 

Furthermore, I found out that `metadata.keywords` is only initialized if the photo has IPTC data. If it's not, my code could not push the keywords to the (not existing) array. So I had to initialize it if not yet done. That's the reason why the keywords of my photo were not present, hence failing the test.

The tests for the photo with special characters was failing because I wrongly used the `.value` instead of `.description` to get the value from the `subject` fields. I found out that I had to use `.description` to get right encoding. However, I still don't understand why the photo has a keyword `squirrel` in both IPTC and XMP but the json file to test against doesn't have it. Nevertheless the test passes. 

To make the test pass, I removed my rating from my test photo. I found out that XnView added to the XMP meta data a value from 0 to 100 as `MicrosoftPhoto.Rating` besides the original tag `xmp.Rating` which goes only from 0 to 5. ExifReader overwrites the second last read value by the last read value. In my case, `MicrosoftPhoto.Rating` overwrote `xmp.Rating` which made the test fail.
I opened an [issue](https://github.com/mattiasw/ExifReader/issues/148) where I suggest to add a new property `RatingPercent` to not overwrite `Rating`.

Last but not least I did not understand why `npm test` failed locally on test `39) SearchQueryParser`, I found out that it has nothing to do with my code changes because this test fails also locally without my code changes. However, the test passes online at github.

Now all tests are green, so my [pull request](https://github.com/bpatrik/pigallery2/pull/270) can be merged into master.
